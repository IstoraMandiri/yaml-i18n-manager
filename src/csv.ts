import { promises as fs } from 'fs';

import csv from 'async-csv';

import type { TranslatedYaml } from './types';

const columns = [
  { key: 'dHash', header: 'dHash' },
  { key: 'path', header: 'Location' },
  { key: 'sig', header: 'Key' },
  { key: 'defaultValue', header: 'Default Language' },
  { key: 'value', header: 'Existing Translation' },
  { key: 'update', header: 'Updated Translation (or leave blank)' },
];

const parseColumns = columns.map(({ key }) => key);

export async function generate(data: TranslatedYaml[]) {
  return await csv.stringify(data, { header: true, columns });
}

export async function parse(path: string) {
  const str = await fs.readFile(path, 'utf8');
  return (await csv.parse(str, { columns: parseColumns, skip_empty_lines: true })).slice(1);
}
