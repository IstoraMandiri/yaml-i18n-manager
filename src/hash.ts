import { createHash } from 'crypto';

export default function (str: string): string {
  return createHash('sha256').update(str, 'utf8').digest('base64').slice(0, 10);
}
