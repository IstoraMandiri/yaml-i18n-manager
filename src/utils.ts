import { createHash } from 'crypto';

import type { ExistingYaml, TranslatedYaml, MatchedYaml } from './types';

export const K_SEP = '__';

export function normalizeSig({ path, sig }: ExistingYaml | TranslatedYaml) {
  const nPath = path.split('.').slice(0, -2).join('.');
  return `${nPath}:${sig.replace(/\[.*?\:/g, '[')}`;
}

export function replaceLocaleInPath(path: string, posftFix: string) {
  return `${path.split('.').slice(0, -2).join('.')}.${posftFix}`;
}

export function hash(str: string): string {
  return createHash('sha256').update(str, 'utf8').digest('base64').slice(0, 10);
}

export function match(yamls: ExistingYaml[]): MatchedYaml[] {
  return yamls.map((dYaml) => {
    const matches = yamls.reduce((o, tYaml) => {
      if (tYaml.locale == dYaml.locale || normalizeSig(tYaml) !== normalizeSig(dYaml)) {
        return o;
      }
      return { [tYaml.locale]: tYaml };
    }, {});
    return { ...dYaml, matches };
  });
}

export function combine(
  matched: MatchedYaml[],
  translated: TranslatedYaml[],
  locale: string,
): TranslatedYaml[] {
  return matched.map(
    (m): TranslatedYaml => {
      const match =
        (m.matches || {})[locale] || translated.find((t) => normalizeSig(t) === normalizeSig(m));
      const path = replaceLocaleInPath(m.path, `${locale}.yaml`);
      return {
        path,
        locale: locale,
        sig: m.sig,
        key: m.key,
        dHash: m.vHash,
        defaultValue: m.value,
        value: match && match.value,
      };
    },
  );
}

export function log(str: string, ...rest: any) {
  console.log(str, ...rest);
}
