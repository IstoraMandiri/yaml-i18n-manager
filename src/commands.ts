import read from './read';
import translate from './translate';
import { combine, match } from './utils';
import { parse } from './csv';
import write from './write';

import type { Config } from './types';

async function getTranslations(targetLocale: string, opts: Config) {
  const { contentDir, filterKeys, defaultLocale } = opts;
  const existing = await read(contentDir);
  const filtered = existing.filter(({ key }) => filterKeys.indexOf(key) === -1);
  const matched = match(filtered).filter(({ locale }) => locale === defaultLocale);
  const requiresTranslating = matched.filter(({ matches }) => !matches[targetLocale]);
  const translated = await translate(requiresTranslating, defaultLocale, targetLocale);
  return combine(matched, translated, targetLocale);
}

const commands: any = {
  // TODO generate-all
  generate: async ({ locale }: { locale: string }, config: Config) => {
    const translations = await getTranslations(locale, config);
    const outputDir = config.outputDir || config.contentDir;
    await write(translations, outputDir);
    console.log(`Wrote ${translations.length} translations to ${outputDir}`);
  },
  export: async ({ locale }: { locale: string }, config: Config) => {
    console.log('export', { locale, config });
  },
  import: async ({ csv }: { csv: string }, config: Config) => {
    const imports = await parse(csv);
    const outputDir = config.outputDir || config.contentDir;
    await write(imports, outputDir);
    console.log(`Wrote ${imports.length} imports to ${outputDir}`);
  },
  // TODO update
};

export default commands;
