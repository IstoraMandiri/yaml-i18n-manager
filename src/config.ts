import yaml from 'yaml';
import { promises } from 'fs';
import { resolve } from 'path';
import { log } from './utils';

const { writeFile, readFile, mkdir } = promises;

export const defaultConfig = {
  defaultLocale: 'en',
  contentDir: 'content',
  yamlDir: '_i18n',
  csvDir: 'csv',
  filterKeys: ['link', 'className', 'date', 'tags'],
  filterMarkdown: [],
};

export async function initConfig(argv: any) {
  const [command] = argv._;
  const location = resolve(process.env.PWD, argv.config);
  if (command === 'init') {
    try {
      const path = location.split('/').slice(0, -1).join('/');
      await mkdir(path, { recursive: true });
      await writeFile(location, yaml.stringify(defaultConfig));
      log(`Wrote config to ${argv.config}`, defaultConfig);
      return null;
    } catch (e) {
      throw e;
    }
  }
  let configData, config;
  try {
    configData = await readFile(location, 'utf8');
  } catch (e) {
    log(location);
    log(
      'Config file not found. Pass a config file with `--config` or initialize with `init` command.',
    );
    return null;
  }
  try {
    config = yaml.parse(configData);
  } catch (e) {
    log(`Could not parse the config file at ${location}`);
    return null;
  }
  return config;
}
