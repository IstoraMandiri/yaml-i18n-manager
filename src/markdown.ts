import yaml from 'yaml';
import fs from 'fs-extra';
import { resolve } from 'path';
import { Config } from './types';
import { scanDir } from './utils';
import { translateYaml } from './translate';

async function parseMarkdonwFile(file: string): Promise<{ header?: any; body: string }> {
  const data = await fs.readFile(file, 'utf8');
  let header, body;
  if (data.startsWith('---')) {
    const [_, _header, _body] = data.split('---');
    body = _body;
    header = yaml.parse(_header);
  } else {
    body = data;
  }
  return { header, body };
}

function normalize(n: string): string {
  const str = n.split('.').slice(0, -2).join('.');
  console.log(str);
  return str;
}

export async function generateMarkdown(locale: string, opts: Config) {
  const markdownFiles = (await scanDir(opts.contentDir))
    .filter(({ ext }) => ['md', 'mdx'].includes(ext))
    .filter(({ relative }) => !opts.filterMarkdown.find((p) => relative.startsWith(p)))
    .map((md) => ({ ...md, locale: md.name.split('.').slice(-2)[0] }))
    .filter(({ locale }) => [opts.defaultLocale, locale].includes(locale));
  const deduped = markdownFiles.filter(
    ({ relative, locale }) =>
      locale === opts.defaultLocale &&
      !markdownFiles.find(
        ({ relative: r2, locale: l2 }) => l2 !== locale && normalize(r2) === normalize(relative),
      ),
  );
  const contents = await Promise.all(
    deduped.map(async (file) => ({ ...file, data: await parseMarkdonwFile(file.path) })),
  );
  const normalized = contents.reduce((a, i): any => {
    const {
      data: { header, body },
      relative,
    } = i;
    const items = [{ relative, key: 'body', value: body }];
    Object.keys(header)
      .filter((k) => !opts.filterKeys.includes(k))
      .forEach((k) => {
        items.push({ relative, key: `header.${k}`, value: header[k] });
      });
    return [...a, ...items];
  }, []);

  const translated = await translateYaml(normalized, opts.defaultLocale, locale);
  const constructed = translated.reduce((o: any, { relative, key, value }: any) => {
    const [k, subKey] = key.split('.');
    const update = o[relative] || {};
    update[k] = subKey ? { ...update[k], [subKey]: value } : value;
    return { ...o, [relative]: update };
  }, {});
  // now write the files.
  await Promise.all(
    Object.keys(constructed).map(async (k) => {
      const newFileName = `${k.split('.').slice(0, -2).join('.')}.${locale}.${k.split('.').pop()}`;
      const { header, body } = constructed[k];
      const data = `---\n${yaml.stringify(header).trim()}\n---${body}`;
      await fs.writeFile(resolve(opts.contentDir, newFileName), data);
    }),
  );
}
