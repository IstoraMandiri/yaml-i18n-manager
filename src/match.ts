import { normalizeSig } from './utils';

export default function match(yamls: ExistingYaml[]): ExistingYaml[] {
  return yamls.map((dYaml) => {
    const matches = yamls.reduce((o, tYaml) => {
      if (tYaml.locale == dYaml.locale || normalizeSig(tYaml) !== normalizeSig(dYaml)) {
        return o;
      }
      return { [tYaml.locale]: tYaml };
    }, {});
    return { ...dYaml, matches };
  });
}
