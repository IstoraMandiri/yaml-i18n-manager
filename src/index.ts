#!/usr/bin/env node
import yargs from 'yargs';
import { defaultConfig, initConfig } from './config';
import commands from './commands';

function requireLocale(y: any) {
  y.positional('locale', {
    describe: 'Target language code (e.g. ja)',
    type: 'string',
  });
}

const argv: any = yargs
  .usage('Usage: $0 <command> [options]')
  .version('v')
  .alias('v', 'version')
  .help('h')
  .alias('h', 'help')
  .options({
    c: {
      alias: 'config',
      type: 'string',
      describe: 'Config file',
      default: `${defaultConfig.managementDir}/config.yaml`,
    },
  })
  .command('init', 'Generate Config File')
  .command('generate <locale>', 'Auto-translate content', requireLocale)
  .command('export <locale>', 'Export CSV for a locale', requireLocale)
  .command('import <csv>', 'Import CSV translations', (y: any) => {
    y.positional('csv', {
      describe: 'Relative CSV file path',
      type: 'string',
    });
  })
  .demandCommand(1, 'Please enter a command').argv;

(async () => {
  const config = await initConfig(argv);
  if (!config) {
    return;
  }
  const [command] = argv._;
  await commands[command](argv, config);
})();
