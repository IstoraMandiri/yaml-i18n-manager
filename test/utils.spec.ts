import { hash, match } from '../src/utils';

describe('utils', () => {
  describe('hash', () => {
    it('hashes', async () => {
      expect(hash('hello')).toMatchSnapshot();
    });
  });

  describe('match', () => {
    it('matches sibling translations', () => {
      expect(
        match(
          [
            { path: 'a.en.yaml', sig: 'a', locale: 'en' },
            { path: 'a.jp.yaml', sig: 'a', locale: 'jp' },
            { path: 'b.en.yaml', sig: 'a[0:test].thing', locale: 'en' },
            { path: 'b.jp.yaml', sig: 'a[0:test].thing', locale: 'jp' },
          ].map((item, i) => ({ ...item, value: 'a', key: 'b', vHash: 'c' })),
        ),
      ).toMatchSnapshot();
    });
  });
});
