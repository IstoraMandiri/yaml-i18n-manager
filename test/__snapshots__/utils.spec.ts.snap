// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils hash hashes 1`] = `"LPJNul+wow"`;

exports[`utils match matches sibling translations 1`] = `
Array [
  Object {
    "key": "b",
    "locale": "en",
    "matches": Object {
      "jp": Object {
        "key": "b",
        "locale": "jp",
        "path": "a.jp.yaml",
        "sig": "a",
        "vHash": "c",
        "value": "a",
      },
    },
    "path": "a.en.yaml",
    "sig": "a",
    "vHash": "c",
    "value": "a",
  },
  Object {
    "key": "b",
    "locale": "jp",
    "matches": Object {
      "en": Object {
        "key": "b",
        "locale": "en",
        "path": "a.en.yaml",
        "sig": "a",
        "vHash": "c",
        "value": "a",
      },
    },
    "path": "a.jp.yaml",
    "sig": "a",
    "vHash": "c",
    "value": "a",
  },
  Object {
    "key": "b",
    "locale": "en",
    "matches": Object {
      "jp": Object {
        "key": "b",
        "locale": "jp",
        "path": "b.jp.yaml",
        "sig": "a[0:test].thing",
        "vHash": "c",
        "value": "a",
      },
    },
    "path": "b.en.yaml",
    "sig": "a[0:test].thing",
    "vHash": "c",
    "value": "a",
  },
  Object {
    "key": "b",
    "locale": "jp",
    "matches": Object {
      "en": Object {
        "key": "b",
        "locale": "en",
        "path": "b.en.yaml",
        "sig": "a[0:test].thing",
        "vHash": "c",
        "value": "a",
      },
    },
    "path": "b.jp.yaml",
    "sig": "a[0:test].thing",
    "vHash": "c",
    "value": "a",
  },
]
`;
